### PR TITLE
[mimictts] Initial contribution

### PR DESCRIPTION
**Pull request by dalgwen**
Add a voice service for Mimic (version 3+) TTS from Mycroft AI. 

# Description

Mimic (version 3) is the brand new Text-To-Speech software designed for the Mycroft vocal assistant.
It is a great improvement from the older versions.
Mimic can be used as a standalone server, and as such it exposes an API that this TTS service makes use of.
This bundle is NOT a standalone, it requires Mimic to run somewhere on your computer/network

[Community discussion on the marketplace](https://community.openhab.org/t/mimic-text-to-speech/137040)

### CODE REVIEW

**Comment by lolodomo** [comment]
I will certainly review. 
And I am hurry to check how the French voice renders.

**Comment by dalgwen** [comment]
One voice (sisiph if I remember correctly) is (really) good, the others are unusable.
With sisiph, there are certainly misprononciation (with silent letter) but words can be rewrited to avoid this. It is not perfect, but imho the best in the offline scene.

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties:10]: Please add this at the end of this file:
```

# service

service.voice.mimictts.label = Mimic Text-to-Speech
```

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java:60]: maybe a private method ?

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java:67]: Maybe you could build a better label ? Maybe with the speaker surrounding by parenthesis ?
```
return name + ((speaker != null) ? " (" + speaker + ")" : "");
```

**Review by lolodomo — CHANGES_REQUESTED** [review]
Review part 1 of 2 (everything except the main class)

**Comment by lolodomo** [comment]
And please add an entry in the file CODEOWNERS.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java
Comment (lolodomo) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java:67]: What is especially expected in the label is to find the language of the voice. So please add it.
In general the format is "name (language)" but you are free to add the speaker if available.

Comment (lolodomo) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java:67]: I removed my wrong comment about adding language in the label, not required. The console command is displaying the language of each voice but using a property of the voice, not its label.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:98]: You do not need to reference ConfigurationAdmin and do what you do in the body of the method.
Method `activate` will be called with a map. Look at Polly TTS or VoiceRSS as examples.

In fact, this method can be removed if you initialize `config` in its declaration.
```
private final MimicConfiguration config = new MimicConfiguration();
```

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:137]: You could catch parsing errors.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:143]: You could catch parsing errors.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:149]: You could catch parsing errors.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:188]: Please use WARN level.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:187]: Please also consider JsonSyntaxException

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:170]: Please use WARN level.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:174]: Please use DEBUG level.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:214]: Can't be null

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:210]: You are not checking that the voice is valid, meaning an available Mimic voice.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Review part 2 of 2

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java:52]: Replacing by `_` would allow more readable UID especially when technical name is `name#speaker` ?

File: bundles/org.openhab.voice.mimictts/README.md
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/README.md:7]: Maybe you can add a link to the documentation explaining how to install this WEB server and the different voices.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.voice.mimictts/README.md
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/README.md]: Done

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/config/config.xml
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/config/config.xml]: Yes ! I copy/paste from another bundle, I should have thinked twice.

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/config/config.xml
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/config/config.xml]: Done

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/config/config.xml
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/config/config.xml]: Done

File: bundles/org.openhab.voice.mimictts/README.md
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/README.md]: Done

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties:10]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java]: Oops I forgot to label it as public.
It is needed in the synthetize method (as a parameter in the URL)

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicVoice.java]: Goog idead, done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Good to know, thanks. Done and tested OK.

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java]: Done

File: bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/java/org/openhab/voice/mimic/internal/MimicTTSService.java:217]: Indeed. I added a check.

File: bundles/org.openhab.voice.mimictts/README.md
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/README.md]: Yes, good idea, I did some rewrite and include it.

**Comment by dalgwen** [comment]
Thank you very much for your review.

In addition to the comments you made, I also made a little rewrite to scan the available voices only at start/reactivate.
I just realized that the method getAvailableVoices() is called before every synthetize() call, and I think we should avoid the related HTTP request as much as possible.

Additionally, I wanted to discuss something else : 
I didn't include any cache mecanism.
I know some other STT bundle do it.
But I think it could implemented at the voice manager level, to avoid many reimplementations by the STT service.
WDYT ? I could maybe make a pull request.

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties:13]: Note that the translations are expected to be done in Crowdin. But ok, I will upload these translations in Crowdin once the PR is merged.
For consistency, please translate rather like that: Synthèse vocale Mimic

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties:8]: `et` and not `and`

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties:4]: To be removed

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties:4]: To be removed.

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Comment by lolodomo** [comment]
> In addition to the comments you made, I also made a little rewrite to scan the available voices only at start/reactivate.
> I just realized that the method getAvailableVoices() is called before every synthetize() call, and I think we should avoid the related HTTP request as much as possible.

You're right and I add in my TODO list to check that all TTS services are doing that property,

**Comment by lolodomo** [comment]
> Additionally, I wanted to discuss something else :
> I didn't include any cache mecanism.
> I know some other STT bundle do it.
> But I think it could be implemented at the voice manager level, to avoid many reimplementations by the different STT services.
> WDYT ? I could maybe make a pull request.

It looks like a good idea.
Please implement a solution that could replace all current implementations in existing voice services.

File: bundles/org.openhab.voice.mimictts/README.md
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/README.md:9]: makes

File: bundles/org.openhab.voice.mimictts/README.md
Comment (lolodomo) [inline bundles/org.openhab.voice.mimictts/README.md:48]: A real voice example would be appreciated (real ID).

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties]: Oh OK, I wasn't aware of that. I thought there was some kind of sync with Crowdin.

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts_fr.properties]: Oops, done.

File: bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/src/main/resources/OH-INF/i18n/mimictts.properties]: Done

File: bundles/org.openhab.voice.mimictts/README.md
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/README.md]: Done

File: bundles/org.openhab.voice.mimictts/README.md
Comment (dalgwen) [reply bundles/org.openhab.voice.mimictts/README.md]: Done

**Review by lolodomo — APPROVED** [review]
LGTM, thank you for your contribution

**Comment by lolodomo** [comment]
Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

**Comment by lolodomo** [comment]
I have uploaded the French translations in Crowdin (while someone was already translating them...).

**Comment by dalgwen** [comment]
Great, thank you !

**Pull request merged**
